text/3470-crabi-v1.md

```diff
@@ -0,0 +1,661 @@
+often uses representations involving structures passed or returned by value.
+This has the net effect of grouping these together into multiple types, without
+introducing unnecessary indirection.
+
+Some crABI representations use an underlying C type but do not use the full
+value space of that type. In that case, passing or returning values outside the
+valid range invokes undefined behavior. Programs may assume that no values
+outside that range are passed or returned, and in particular the compiler may
+generate code that does not check this assumption, or may optionally include
+validation assertions when debugging.
+
+## `char` - Unicode character type
+
+crABI supports Unicode characters, which in Rust use the `char` type. This
+translates to the C `uint32_t` type. This type will never contain a value
+larger than 0x10FFFF, or a value in the range 0xD800 to 0xDFFF inclusive.
+
+Note that there is no special handling for an array of values of this type,
+which is not equivalent to a string (unless using a UCS-4 encoding).
+
+## Zero-sized types (e.g. `()` and `PhantomData`)
+
+crABI supports zero-sized types, such as `()` and `PhantomData`, as long as
+they have an alignment of 1. Thse types have only one value, and require zero
+bits to convey, so they are not passed through the ABI at all, and do not
+appear in the translated C types or functions.
+
+Pointers and references to zero-sized types are valid in crABI, and are passed
+like any other pointer or reference to an opaque Rust type.
+
+## `&T` and `&mut T` - borrowed reference
+
+As with `extern "C"`, crABI supports references (immutable and mutable) to
+types. These do not pass ownership of the value, only a pointer to the value.
+(Note that crABI 1.0 does not have lifetime information; the provider of an API
+must document the lifetime of a reference, and the recipient of the reference
+must handle it accordingly.)
+
+Both types of reference are passed as a pointer to the underlying type. The
+pointer must never be null.
+
+## Tuple
+
+crABI supports tuples with arbitrarily many fields. These are translated to a
+by-value C structure containing fields of the specified types in the same
+order. For example:
+
+```rust
+extern "crabi" fn func(a: u32, t: cr#(u64, u16), b: i8) -> cr#(u8, u32);
+```
+
+is equivalent to:
+
+```c
+struct func_t_arg {
+    uint64_t f1;
+    uint16_t f2;
+};
+struct func_ret {
+    uint8_t f1;
+    uint32_t f2;
+};
+extern struct func_ret func(uint32_t a, struct func_t_arg t, int8_t b);
+```
+
+Note that crABI tuples are not the same type as Rust tuples, as Rust tuples do
+not guarantee a stable layout.
+
+## `&[T]` or `&mut [T]` - borrowed slice
+
+crABI supports slices (references to arrays) as long as crABI supports the
+element type.
+
+Slices translate to a by-value struct containing two fields: a pointer to the
+element type, and a `size_t` number of elements, in that order.
```

Comment:
Does this properly translate to the same limitation as rust slices? I.e. at most isize::MAX bytes but at most usize::MAX items (for ZSTs)?